[CodeCompletion] Migrate argument position completion to the solver-based implementation

ahoppen opened the pull request:
Revised version of #38266 with most stress tester failures fixed.

CC @nathawes 

rdar://83435550

ahoppen commented:
@swift-ci Please SourceKit stress test

hamishknight commented on lib/IDE/CodeCompletion.cpp:
```diff
@@ -6732,7 +6733,142 @@ static void deliverCompletionResults(CodeCompletionContext &CompletionContext,
+
+  ParameterList *PL = nullptr;
+  if (Ret.FuncD) {
+    if (auto *FD = dyn_cast<AbstractFunctionDecl>(Ret.FuncD)) {
+      PL = FD->getParameters();
+    } else if (auto *SD = dyn_cast<SubscriptDecl>(Ret.FuncD)) {
+      PL = SD->getIndices();
+    } else if (auto *EED = dyn_cast<EnumElementDecl>(Ret.FuncD)) {
+      PL = EED->getParameterList();
+    } else {
+      // No parameter list for function-type params/variables.
+      assert(isa<VarDecl>(Ret.FuncD) && "Unhandled decl type?");
+    }
```
Nit: You could use the top-level `swift::getParameterList` here

hamishknight commented on lib/Sema/CSSimplify.cpp:
```diff
@@ -2029,6 +2116,36 @@ static bool fixExtraneousArguments(ConstraintSystem &cs,
+  if (!commonAncestor || commonAncestor == fixExpr)
+    return nullptr;
+
+  if (!commonAncestor || (!isa<CallExpr>(commonAncestor) && !isa<SubscriptExpr>(commonAncestor)))
```
Should this check `commonAncestor->getArgs()` for null instead of matching parent types to e.g also handle ObjectLiteralExprs?

hamishknight commented on lib/Sema/TypeCheckCodeCompletion.cpp:
```diff
@@ -1325,3 +1349,144 @@ void KeyPathTypeCheckCompletionCallback::sawSolution(
 }
+
+static bool isInPatternMatch(CodeCompletionExpr *CE, ConstraintSystem &CS) {
+  if (auto *TE = dyn_cast_or_null<TupleExpr>(CS.getParentExpr(CE))) {
+    if (!TE->isImplicit() || TE->getNumElements() != 2 ||
+        TE->getElement(0) != CE)
+      return false;
+    if (auto *DRE = dyn_cast<DeclRefExpr>(TE->getElement(1))) {
+      auto Name =  DRE->getDecl()->getName();
+      return Name && Name.isSimpleName("$match");
+    }
+  }
```
Does this need updating for the ArgumentList changes?

hamishknight commented on lib/Sema/TypeCheckCodeCompletion.cpp:
```diff
@@ -1325,3 +1349,144 @@ void KeyPathTypeCheckCompletionCallback::sawSolution(
+
+  Expr *ParentCall = CompletionExpr;
+  while (ParentCall &&
+         !(isa<CallExpr>(ParentCall) || isa<SubscriptExpr>(ParentCall) || isa<BinaryExpr>(ParentCall)))
```
Should this also be checking `ParentCall->getArgs()` for null instead of matching parent types?

hamishknight commented on lib/Sema/TypeCheckCodeCompletion.cpp:
```diff
@@ -1325,3 +1349,144 @@ void KeyPathTypeCheckCompletionCallback::sawSolution(
+  bool HasLabel = false;
+  if (auto PE = CS.getParentExpr(CompletionExpr)) {
+    if (auto Args = PE->getArgs()) {
+      HasLabel = !Args->get(ArgIdx).getLabel().empty();
```
Nit: You could use `Args->getLabel(ArgIdx)` here

hamishknight reviewed (commented):
Cool! Hope you don't mind some drive-by nits :)

ahoppen commented:
Thanks @hamishknight for the suggestions for the new `ArgumentList` API. I just rebased it so it worked and didn’t look if the code could be improved.

For now, I’m looking to merge the changes to solution scoring (which are needed to fix the stress tester issues) here: https://github.com/apple/swift/pull/39392

ahoppen commented:
@swift-ci Please SourceKit stress test

ahoppen commented:
@swift-ci Please smoke test

ahoppen commented:
@swift-ci Please SourceKit stress test

ahoppen commented:
@swift-ci Please SourceKit stress test

ahoppen commented:
@swift-ci Please SourceKit stress test

ahoppen commented:
@swift-ci Please smoke test

ahoppen commented:
@swift-ci Please SourceKit stress test

ahoppen committed:
[TypeChecker] Add method to check whether a type variable represents a code completion token

ahoppen commented on lib/Sema/TypeCheckCodeCompletion.cpp:
```diff
@@ -527,8 +556,20 @@ static void filterSolutions(SolutionApplicationTarget &target,
+      // We don't care how much source code was ignored after the code
+      // completion token, just whether if there was any.
+      if (S.getFixedScore().Data[SK_IgnoreAfterCompletionToken] > 0 &&
+          minScore.Data[SK_IgnoreAfterCompletionToken] == 0) {
```
I am not really happy with this part yet. What I am really trying to achieve is to have a boolean flag in the score that indicates whether code after the code completion token was ignored – I don’t care how much.  At the moment we might end up with solutions that have a `SK_IgnoreAfterCompletionToken` score of >1 (sometimes also created by combining two partial solutions, where each partial solution ignored arguments). 

In the current implementation, we might filter out solutions with a `SK_IgnoreAfterCompletionToken` of 2 when comparing against a solution that has a `SK_IgnoreAfterCompletionToken` of 1 in the constraint system, even though we don’t really consider it worse.

Options that I have considered:
- Only increase score in `recordFix` if `SK_IgnoreAfterCompletionToken` is 0 — doesn’t work because we might combine two partial solutions that both have `SK_IgnoreAfterCompletionToken = 1`
- Hacking the `operator<` of `Score` to consider all `SK_IgnoreAfterCompletionToken > 0` equally bad — this seems very surprising to me
- Hack the `operator+` of `Score` such that `1 + 1 = 1` for `SK_IgnoreAfterCompletionToken` — this makes `Score` addition non-linear, not sure what effects it might have
- Leave it as-is for now because we don’t have any counter-examples that break yet

xedin commented on lib/Sema/CSSimplify.cpp:
```diff
@@ -7631,14 +7783,14 @@ static bool isForKeyPathSubscript(ConstraintSystem &cs,
 
-static bool isForKeyPathSubscriptWithoutLabel(ConstraintSystem &cs,
-                                              ConstraintLocator *locator) {
+static bool mayBeForKeyPathSubscriptWithoutLabel(ConstraintSystem &cs,
```
Nit: lower-case `b` here.

ahoppen replied:
I think the capital `b` is correct here as it’s “may be for key path subscript without label” where “be” is a verb.

xedin replied:
Okay, makes sense, `<locator> may be for ...` definitely makes sense.

ahoppen commented:
@swift-ci Please smoke test

ahoppen commented:
After some discussion with @xedin, we have decided to always ignore arguments after the code completion token because it defines away an entire family of ambiguities where it’s unclear whether these arguments should influence what’s suggested for the code completion token because we can’t know whether the user wants to rewrite the entire call site or only update one parameter. We might be able to revisit this decision in a follow-up PR. As a side-effect, it also simplifies this PR.

ahoppen commented:
@swift-ci Please smoke test

ahoppen commented:
@swift-ci Please SourceKit stress test

ahoppen commented:
@swift-ci Please smoke test

ahoppen commented:
@swift-ci Please SourceKit stress test

xedin commented on lib/Sema/CSGen.cpp:
```diff
@@ -747,7 +747,73 @@ namespace {
+                                   TVO_CanBindToNoEscape | TVO_CanBindToHole);
+    };
+
+    if (auto closure = dyn_cast<ClosureExpr>(expr)) {
```
There is a method for this already - `inferClosureType` which should do the right thing since parameters are marked as TVO_CanBindToHole by default.

ahoppen replied:
Ah, nice. I needed to add another parameter to `inferClosureType` so that the result can also bind to a hole when ignoring arguments, but it’s definitely nicer than my hand-rolled implementation.

xedin commented on lib/Sema/CSSimplify.cpp:
```diff
@@ -1213,6 +1234,68 @@ class ArgumentFailureTracker : public MatchCallArgumentListener {
+    // (causing the solution to have a worse score) as the user just hasn't
+    // written them yet. Early exit to avoid recording them in this case.
+    if (ArgInfo.allowsMissingArgAt(argInsertIdx, param)) {
+      unsigned newArgIdx = Arguments.size();
```
I think it's worth extracting this whole body of `if` into a separate method on a `ArgumentFailureTracker`, so it could be used by both versions of `missingArgument`.

ahoppen replied:
Good idea 👍

xedin reviewed (approved):
Solver related changes look good!

ahoppen committed:
[CodeCompletion][Sema] Migrate CallArgurment position completion to the solver-based implementation

This hooks up call argument position completion to the typeCheckForCodeCompletion API to generate completions from all the solutions the constraint solver produces (even those requiring fixes), rather than relying on a single solution being applied to the AST (if any).

Co-authored-by: Nathan Hawes <nathan.john.hawes@gmail.com>

ahoppen commented:
@swift-ci Please smoke test

ahoppen commented:
I addressed your review comments @xedin and made the following changes:
- Don’t optimize constraints of arguments that are being ignored: [CSGen.cpp:791](https://github.com/apple/swift/pull/39373/files#diff-5fffd7ce67aca13059738c34fbc714dcbaf3f6fe83daff7ca75836bc88304c20R791-R794)
- Extracted logic whether a argument should be claimed despite label mismatch (because it is the completion expression) to `CompletionArgumentTracker` ([CSSimplify.cpp:1295](https://github.com/apple/swift/pull/39373/files#diff-f7934d7c1cacc858eca64d34e090519bec4b0c6c3cd15ffe8fe92013eb65a6ccR1295-R1304)). This makes sure that we only claim arguments despite label mismatches when we are performing solver-based completion. It broke a couple of test cases in the legacy code completion.

xedin reviewed (approved):
LGTM! We have discussed this offline but I'm just going to leave it here - the storage for ignored arguments would be switched to use `Expr *` and check in ConstraintWalker @ [CSGen.cpp:791](https://github.com/apple/swift/pull/39373/files#diff-5fffd7ce67aca13059738c34fbc714dcbaf3f6fe83daff7ca75836bc88304c20R791-R794) would have to get wrapped into `if (cs.isForCodeCompletion())` in a follow-up.
